### PR TITLE
DYN-5018 wires disappear when moving group

### PR DESCRIPTION
**Pull request by filipeotero**
### Purpose

In this PR, the position of the wires is calculated based on the height of the port that is created after the group is collapsed.
Previously, we were assuming that all the ports have the same height, causing a mispositioning of the wires. 

image

image

Positionwires

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@QilongTang @pinzart @reddyashish

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 1]
```diff
@@ -260,6 +262,15 @@ public bool IsExpanded
                 AddGroupToGroupCommand.RaiseCanExecuteChanged();
                 RaisePropertyChanged(nameof(IsExpanded));
                 RedrawConnectors();
+                ReportNodesPosition();
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:265]: `node.ReportPosition` is an expensive operation when dealing with a lot of nodes under the same group.
Please check the performance with (there should be a .dyn you can use in this [jira taks](https://jira.autodesk.com/browse/DYN-4600)

Also take a look at what code paths might be calling the IsExpanded setter (to get a better idea what to test) - could be direct function calls or RaisePropertyChanges...

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 2]
```diff
@@ -746,15 +752,19 @@ private List<PortViewModel> CreateProxyInPorts(IEnumerable<PortModel> groupPortM
...
-            for (int i = 0; i < groupPortModels.Count(); i++)
+            double verticalPosition = 0;
+            foreach (var group in groupPortModels)
             {
-                var model = groupPortModels.ElementAt(i);
-                newPortViewModels.Add(originalPortViewModels[i].CreateProxyPortViewModel(model));
-
-                // calculate new position for the proxy inports.
-                model.Center = CalculatePortPosition(model, i);
+                var originalPort = originalPortViewModels.FirstOrDefault(x => x.PortModel.GUID == group.GUID);
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:758]: I'm, confused by then, but not super familiar with this code, when would the entire guid of a group equal the guid of a port model?

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 3]
```diff
@@ -245,9 +245,11 @@ public bool IsExpanded
             set
             {
                 annotationModel.IsExpanded = value;
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:247]: I think a better option would be to listen to the model.IsExpanded OnPropertyChanged...instead of the entire setter
(this can be done in another PR) 

annotationModel.PropertyChanged += OnModelPropertyChanged;
OnModelPropertyChanged(sender, e) {
   if (e.PropertyName == "IsExpanded") {
                      InPorts.Clear();
                OutPorts.Clear();
                if (value)
                {
                    this.ShowGroupContents();       
                }
                else
                {
                    this.SetGroupInputPorts();
                    this.SetGroupOutPorts();
                    this.CollapseGroupContents(true);
                    RaisePropertyChanged(nameof(NodeContentCount));
                }
                WorkspaceViewModel.HasUnsavedChanges = true;
                .............the whole setter content............
   }
}

ANd only keep the getter

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 4]
```diff
@@ -1081,10 +1097,12 @@ private void model_PropertyChanged(object sender, System.ComponentModel.Property
                     RaisePropertyChanged("Width");
                     RaisePropertyChanged(nameof(ModelAreaRect));
                     UpdateAllGroupedGroups();
+                    UpdateProxyPortsPosition();
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:1100]: check performance here...
When moving a group (drag with mouse) ...these properties can get changed a lot of times

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 1 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:265]: It is for when the group is uncollapsed and the annotation inside it needs to report the position. Otherwise, that's what happens:
annotationNode

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 2 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:758]: It's the group's port. I will change the variable to groupPort to avoid confusion. Thanks.

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 3 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:247]: Ok, let's do that on another PR

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 4 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:1100]: Ok, I will check an alternative for this.

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 1 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:265]: If there is no performance penalty in doing "node.ReportPosition" then it is fine.
The fix would be to make the Connectors redraw when the corresponding NodeModels(start or end) change their IsProxyPort flag
https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCoreWpf/ViewModels/Core/ConnectorViewModel.cs#L1209

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 4 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:1100]: check if performance is an issue first...then check alternatives
No use in doing premature optimization

**Comment by filipeotero** [comment]
Instead of calling the `UpdateProxyPortsPosition` method every time the width changes, the method is being called on `GroupTextBox_OnTextChanged ` which was updating of the width of the group, but not the center of its ports.

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 5]
```diff
@@ -536,11 +547,15 @@ private bool CanChangeFontSize(object obj)
         public AnnotationViewModel(WorkspaceViewModel workspaceViewModel, AnnotationModel model)
         {
             annotationModel = model;
+            var modelBase = (ModelBase)model;
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:550]: why do you need to cast it to ModelBase ?

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 6]
```diff
@@ -536,11 +547,15 @@ private bool CanChangeFontSize(object obj)
...
+
             this.WorkspaceViewModel = workspaceViewModel;
             this.preferenceSettings = WorkspaceViewModel.DynamoViewModel.PreferenceSettings;
             model.PropertyChanged += model_PropertyChanged;
             model.RemovedFromGroup += OnModelRemovedFromGroup;
             model.AddedToGroup += OnModelAddedToGroup;
+            modelBase.PropertyChanged += ModelBase_PropertyChanged;
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:557]: there already is a handler for the model PropertyChange event
```
model.PropertyChanged += model_PropertyChanged;
```

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 7]
```diff
@@ -625,7 +649,6 @@ internal void SetGroupInputPorts()
         /// </summary>
         internal void SetGroupOutPorts()
         {
-            OutPorts.Clear();
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:628]: why did you remove `OutPorts.Clear();` ?

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 8]
```diff
@@ -581,14 +596,23 @@ public AnnotationViewModel(WorkspaceViewModel workspaceViewModel, AnnotationMode
...
-        /// Input ports that either is connected to a Node outside of the
-        /// group, or has a port that is not connected will be used for the group.
-        /// </summary>
+            /// Creates input ports for the group based on its Nodes.
+            /// Input ports that either is connected to a Node outside of the
+            /// group, or has a port that is not connected will be used for the group.
+            /// </summary>
         internal void SetGroupInputPorts()
         {
-            InPorts.Clear();
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:591]: why did you remove `InPorts.Clear();` ?

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 5 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs]: Removed

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 9]
```diff
@@ -581,14 +596,23 @@ public AnnotationViewModel(WorkspaceViewModel workspaceViewModel, AnnotationMode
             LoadGroupStylesFromPreferences(preferenceSettings.GroupStyleItemsList);
         }
 
+        private void ModelBase_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            switch (e.PropertyName)
+            {
+                case nameof(annotationModel.Position):
+                    UpdateProxyPortsPosition();
```
[the remark this answers is not in the document]
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs]: removed

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 8 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:591]: Same here

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs [hunk 10]
```diff
@@ -581,14 +594,14 @@ public AnnotationViewModel(WorkspaceViewModel workspaceViewModel, AnnotationMode
...
+        
         /// <summary>
-        /// Creates input ports for the group based on its Nodes.
-        /// Input ports that either is connected to a Node outside of the
-        /// group, or has a port that is not connected will be used for the group.
-        /// </summary>
+            /// Creates input ports for the group based on its Nodes.
+            /// Input ports that either is connected to a Node outside of the
+            /// group, or has a port that is not connected will be used for the group.
+            /// </summary>
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:602]: looks like the formatting got screwed up

File: src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs (hunk 7 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:628]: so now `SetGroupInputPorts` and `SetGroupOutPorts` will always append ports to `OutPorts` and `InPorts`.  I hope nobody uses these functions expecting the port arrays to be `re-set` ...as the func names suggest. Can you make these functions private and maybe add to the description that they always append to the respective port arrays ?

Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs:628]: Sure, editted.

**Review by pinzart90 — APPROVED** [review]

**Review by QilongTang — APPROVED** [review]
